drivers/entropy/entropy_esp32.c

```diff
@@ -8,52 +8,77 @@
 
 #include <string.h>
 #include <drivers/entropy.h>
+#include <soc/dport_reg.h>
+#include <soc/rtc.h>
+#include <soc/wdev_reg.h>
+#include <soc/rtc_cntl_reg.h>
+#include <soc/apb_ctrl_reg.h>
+#include <esp_system.h>
+#include <soc.h>
+
+extern int esp_clk_cpu_freq(void);
+extern int esp_clk_apb_freq(void);
 
 static inline uint32_t entropy_esp32_get_u32(void)
 {
-	/*
-	 * APB Address:   0x60035144 (Safe,slower writes)
-	 * DPORT Address: 0x3ff75144 (write bug, fast writes)
-	 * In this case it won't make a difference because it is read only
-	 * More info available at:
-	 * https://www.esp32.com/viewtopic.php?f=2&t=3033&p=14227
-	 * also check: ECO and Workarounds for Bugs Document, point 3.3
+	/* The PRNG which implements WDEV_RANDOM register gets 2 bits
+	 * of extra entropy from a hardware randomness source every APB clock cycle
+	 * (provided WiFi or BT are enabled). To make sure entropy is not drained
+	 * faster than it is added, this function needs to wait for at least 16 APB
+	 * clock cycles after reading previous word. This implementation may actually
+	 * wait a bit longer due to extra time spent in arithmetic and branch statements.
+	 *
+	 * As a (probably unncessary) precaution to avoid returning the
+	 * RNG state as-is, the result is XORed with additional
+	 * WDEV_RND_REG reads while waiting.
 	 */
-	volatile uint32_t *rng_data_reg = (uint32_t *)DT_INST_REG_ADDR(0);
 
-	/* Read just once.  This is not optimal as the generator has
-	 * limited throughput due to scarce sources of entropy, specially
-	 * with the radios turned off.  Might want to revisit this.
+	/* This code does not run in a critical section, so CPU frequency switch may
+	 * happens while this code runs (this will not happen in the current
+	 * implementation, but possible in the future). However if that happens,
+	 * the number of cycles spent on frequency switching will certainly be more
+	 * than the number of cycles we need to wait here.
 	 */
-	return *rng_data_reg;
+	uint32_t cpu_to_apb_freq_ratio =
+		esp_clk_cpu_freq() / esp_clk_apb_freq();
+
+	static uint32_t last_ccount;
+	uint32_t ccount;
+	uint32_t result = 0;
+
+	do {
+		ccount = XTHAL_GET_CCOUNT();
+		result ^= REG_READ(WDEV_RND_REG);
```

Comment:
This is going off on a tangent, but I'm not sure I trust this.

Can this loop ever execute only once?  E.g. if it's been a long time since the last time it was invoked, so the cycle difference is large?

And if it does, is it possible that `WDEV_RND_REG` might not be changed between the XOR in the loop and the XOR on exit?

If so, this function would seem to return zero more often than expected.

Reply:
> Can this loop ever execute only once? E.g. if it's been a long time since the last time it was invoked, so the cycle difference is large?

Yes, it will execute once most of the time where current call timestamp is 16 APB clocks longer than last call. Otherwise, it loops a few times.

> And if it does, is it possible that WDEV_RND_REG might not be changed between the XOR in the loop and the XOR on exit?

WDEV_RND_REG shall not return same value on subsequent readings, not causing the zero result output.

Reply:
If the register is completely different each time it's read, but you just need the delay to make sure the required entropy has been incorporated, this logic would seem to be more clear:
```
	static uint32_t last_ccount;
	uint32_t ccount;
	uint32_t result;

	do {
		ccount = XTHAL_GET_CCOUNT();
		result = REG_READ(WDEV_RND_REG);
	} while (ccount - last_ccount < cpu_to_apb_freq_ratio * 16);

	last_ccount = ccount;
	return result;
```
I get concerned about comments like "we'll xor the value but it's probably not necessary", as well as the repeated XORing if the value is supposed to be 32-bits of random. Extra manipulation doesn't seem to increase the randomness.

Reply:
@pabigot, this implementation came from esp-idf and has a background history (...). After checking internally with the team, it makes sense to dismiss the extra manipulation as you said. Thanks.